Apply more constexpr for offsetOf

Constellation opened the pull request:
#### 51b187d9b056f1f00e506ec24d581867db2f24e5
<pre>
Apply more constexpr for offsetOf
<a href="https://bugs.webkit.org/show_bug.cgi?id=275512">https://bugs.webkit.org/show_bug.cgi?id=275512</a>
<a href="https://rdar.apple.com/129874627">rdar://129874627</a>

Reviewed by Yijia Huang.

More expansion of constexpr OBJECT_OFFSETOF. And changing some more ASSERT with static_assert.

* Source/JavaScriptCore/bytecode/InternalFunctionAllocationProfile.h:
(JSC::InternalFunctionAllocationProfile::offsetOfStructureID):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::linkOSRExits):
* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::offsetOfScrambledBits):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::loadMegamorphicProperty):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/FunctionRareData.h:
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSCallee.h:
(JSC::JSCallee::offsetOfScopeChain):
* Source/JavaScriptCore/runtime/JSFunction.h:
(JSC::JSFunction::offsetOfExecutableOrRareData):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::offsetOfInlineStorage):
(JSC::indexRelativeToBase):
* Source/JavaScriptCore/runtime/NativeExecutable.h:
* Source/WTF/wtf/WeakRandom.h:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::flagsOffset):
(WTF::StringImpl::dataOffset):
(WTF::StringImpl::headerSize):
(WTF::StringImpl::tailOffset):
* Source/WTF/wtf/text/SymbolImpl.h:
(WTF::SymbolImpl::SymbolImpl):

Canonical link: <a href="https://commits.webkit.org/280046@main">https://commits.webkit.org/280046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab5b153325cc91fb7132859549e5c93b4968b568

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55571 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34894 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58555 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6001 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6200 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44753 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/4125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57600 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32797 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47898 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25880 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5227 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4145 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48648 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51514 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60146 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54808 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30725 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5617 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52184 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47968 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51666 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32891 "Built successfully") | | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/67262 "Failed to checkout and rebase branch from PR 29841") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8198 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31557 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/67262 "Failed to checkout and rebase branch from PR 29841") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me

Jarred-Sumner commented on Source/JavaScriptCore/bytecode/InternalFunctionAllocationProfile.h:
```diff
@@ -33,7 +33,7 @@ namespace JSC {
 class InternalFunctionAllocationProfile {
 public:
-    static inline ptrdiff_t offsetOfStructureID() { return OBJECT_OFFSETOF(InternalFunctionAllocationProfile, m_structureID); }
+    static constexpr ptrdiff_t offsetOfStructureID() { return OBJECT_OFFSETOF(InternalFunctionAllocationProfile, m_structureID); }
```
I wonder if `consteval` would be good in these cases

Constellation committed:
Apply more constexpr for offsetOf
https://bugs.webkit.org/show_bug.cgi?id=275512
rdar://129874627

Reviewed by Yijia Huang.

More expansion of constexpr OBJECT_OFFSETOF. And changing some more ASSERT with static_assert.

* Source/JavaScriptCore/bytecode/InternalFunctionAllocationProfile.h:
(JSC::InternalFunctionAllocationProfile::offsetOfStructureID):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::linkOSRExits):
* Source/JavaScriptCore/heap/FreeList.h:
(JSC::FreeCell::offsetOfScrambledBits):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::loadMegamorphicProperty):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/FunctionRareData.h:
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSCallee.h:
(JSC::JSCallee::offsetOfScopeChain):
* Source/JavaScriptCore/runtime/JSFunction.h:
(JSC::JSFunction::offsetOfExecutableOrRareData):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::offsetOfInlineStorage):
(JSC::indexRelativeToBase):
* Source/JavaScriptCore/runtime/NativeExecutable.h:
* Source/WTF/wtf/WeakRandom.h:
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::flagsOffset):
(WTF::StringImpl::dataOffset):
(WTF::StringImpl::headerSize):
(WTF::StringImpl::tailOffset):
* Source/WTF/wtf/text/SymbolImpl.h:
(WTF::SymbolImpl::SymbolImpl):

Canonical link: https://commits.webkit.org/280046@main
